### PR TITLE
chore: upstream rfl and symm tactics

### PR DESCRIPTION
**Pull request by semorrison**
The `trans` tactic in Mathlib needs more polish than these, so I'll do that in a separate PR to ease reviewing.

### CODE REVIEW

File: Std/Lean/Elab/Tactic.lean [hunk 1]
```diff
@@ -0,0 +1,18 @@
+/-
+Copyright (c) 2022 Mario Carneiro. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Mario Carneiro
+-/
+import Lean
```
Comment (digama0) [inline Std/Lean/Elab/Tactic.lean:6]: minimize import

Comment (semorrison) [reply Std/Lean/Elab/Tactic.lean]: Done, and linter at #259.

**Comment by joehendrix** [comment]
These seem like instances of a potential more general "resolution" tactic that would try to resolve a collection of implications  `A1 -> ... AN -> B` against the goal or a hypothesis.

Is that correct?

**Comment by digama0** [comment]
Yes (you are describing `solve_by_elim`), but that's an incredibly heavy hammer and `rfl` and `symm` need to have more controlled results because they are used as structural tactics (nonterminally)

File: Std/Tactic/Relation/Symm.lean [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+    let .app (.app rel _) _ := targetTy | fail
+    let key ← withReducible <| DiscrTree.mkPath rel
+    symmExt.add (decl, key) kind
+}
+
+end Std.Tactic
+
+open Std.Tactic
+
+namespace Lean.Expr
```
Comment (digama0) [inline Std/Tactic/Relation/Symm.lean:48]: All of these functions should be in the `Std.Tactic` namespace, I don't think it's a good idea to create `Expr.symm` here, this is likely to conflict with uses outside std. Ditto for `MVarId.symm`, we don't have a precedent for doing tactic functions with `MVarId` dot notation in std, only lean core does that and I am concerned about name collisions.

**Comment by semorrison** [comment]
> Yes (you are describing `solve_by_elim`), but that's an incredibly heavy hammer and `rfl` and `symm` need to have more controlled results because they are used as structural tactics (nonterminally)

I'll be upstreaming `solve_by_elim` soon.

File: Std/Tactic/Relation/Symm.lean (hunk 2 above)
Comment (semorrison) [reply Std/Tactic/Relation/Symm.lean:48]: I guess so. It makes me a little sad to lose the dot notation for both `Expr.symm` and `MVarId.symm`.

Comment (digama0) [reply Std/Tactic/Relation/Symm.lean:48]: Is it used ever (outside this file)?

Comment (semorrison) [reply Std/Tactic/Relation/Symm.lean:48]: `MVarId.symm` is used in one tactic in Mathlib, and `MVarId.rfl` is used in two (`rw?` and `gcongr`). So it's no great loss. I just hadn't realised you wanted to be as cautious about the namespaces for Std; given Std is always going to be near the top of any import hierarchy I'd thought we'd be safe to insert things into core namespaces.

Comment (digama0) [reply Std/Tactic/Relation/Symm.lean:48]: I think this is mostly fine, but I would try to limit this to things that are both public API and also not likely to be desired for other purposes either upstream or downstream.

Comment (semorrison) [reply Std/Tactic/Relation/Symm.lean:48]: Okay, I have changed the incorrect `Mathlib` namespaces to `Std`, and made the `symmAux` functions private.

Comment (joehendrix) [reply Std/Tactic/Relation/Symm.lean:48]: I also slightly prefer a longer name that indicates it's applying a tactic for these (`applySymm`?), but I'll defer to the author on this.

Is there any remaining changes desired before merging?

Comment (semorrison) [reply Std/Tactic/Relation/Symm.lean:48]: Indeed, the longer name is better. Done now. I'm happy for these to go in.

File: Std/Tactic/Relation/Rfl.lean [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+/-- `MetaM` version of the `rfl` tactic.
+
+This tactic applies to a goal whose target has the form `x ~ x`, where `~` is a reflexive
+relation, that is, a relation which has a reflexive lemma tagged with the attribute [refl].
+-/
+def _root_.Lean.MVarId.applyRfl (goal : MVarId) : MetaM Unit := do
+  let .app (.app rel _) _ ← whnfR <|← instantiateMVars <|← goal.getType
+    | throwError "reflexivity lemmas only apply to binary relations, not
+      {indentExpr (← goal.getType)}"
```
Comment (digama0) [inline Std/Tactic/Relation/Rfl.lean:52]:
```suggestion
    | throwError "reflexivity lemmas only apply to binary relations, not{
        indentExpr (← goal.getType)}"
```

File: Std/Tactic/Relation/Rfl.lean [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+  let s ← saveState
+  let mut ex? := none
+  for lem in ← (reflExt.getState (← getEnv)).getMatch rel do
+    try
+      let gs ← goal.apply (← mkConstWithFreshMVarLevels lem)
+      if gs.isEmpty then return () else
+        logError <| MessageData.tagged `Tactic.unsolvedGoals <| m!"unsolved goals\n
+          {goalsToMessageData gs}"
```
Comment (digama0) [inline Std/Tactic/Relation/Rfl.lean:60]:
```suggestion
        logError <| MessageData.tagged `Tactic.unsolvedGoals <| m!"unsolved goals\n{
          goalsToMessageData gs}"
```

File: Std/Tactic/Relation/Rfl.lean [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    throw e
+  else
+    throwError "rfl failed, no lemma with @[refl] applies"
+
+/--
+This tactic applies to a goal whose target has the form `x ~ x`, where `~` is a reflexive
+relation, that is, a relation which has a reflexive lemma tagged with the attribute [refl].
+-/
+elab_rules : tactic
+| `(tactic| rfl) => withMainContext do liftMetaFinishingTactic (·.applyRfl)
```
Comment (digama0) [inline Std/Tactic/Relation/Rfl.lean:75]:
```suggestion
  | `(tactic| rfl) => withMainContext do liftMetaFinishingTactic (·.applyRfl)
```

**Pull request merged**